packages/gatsby/src/utils/handle-flags.ts

```diff
@@ -0,0 +1,89 @@
+import _ from "lodash"
+import { isCI } from "gatsby-core-utils"
+import realTerminalLink from "terminal-link"
+import { IFlag } from "./flags"
+
+const terminalLink = (text, url): string => {
+  if (process.env.NODE_ENV === `test`) {
+    return `${text} (${url})`
+  } else {
+    return realTerminalLink(text, url)
+  }
+}
+
+const handleFlags = (
+  flags: Array<IFlag>,
+  configFlags: Array<Array<string>>,
+  executingCommand = process.env.gatsby_executing_command
+): { validConfigFlags: Array<IFlag>; message: string } => {
+  // Prepare config flags.
+  // Filter out any flags that are set to false.
+  const filteredConfigFlags = _.toPairs(configFlags).filter(pair => pair[1])
+
+  //  validate flags against current ones
+  //  - any that don't exist — remove silently
+  let validConfigFlags = flags.filter(ae =>
+    filteredConfigFlags.some(flagPair => flagPair[0] === ae.name)
+  )
```

Comment:
This has a lot of small recursion can we make it smaller?

If we convert the flags to a Map it's easier to get things from.

```suggestion
   const availableFlags = new Map();
   flags.forEach(flag => availableFlags.set(flag.name, flag));
   const enabledConfigFlags = Object.keys(configFlags).filter(name => configFlags[name] && availableFlags.has(name))
```